Create 'add account' dialog

mankyKitty opened the pull request:
This PR adds the dialog for creating a wallet only account, with the inactive button for creating vanity accounts.

Commit:
WIP create addaccount dialog

Commit:
Add account dialog added to wallet screen

Creates an account in the wallet only, using the public key as the account name and shows
confirmation dialog.

Commit:
Clear some warnings and remove redundant inputs

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AddAccount.hs:
```diff
@@ -0,0 +1,93 @@
+
+  modalFooter $ do
+    onCancel <- cancelButton def "Cancel"
+    text " "
```
Placeholder for margin?

mankyKitty replied:
Copy-pasta from the modal I was building my implementation from, I suspect. But yes, a placeholder for a margin that should be done in CSS. I'll remove it.

tomsmalley replied:
Since this is pre-modal redesign, perhaps we should remove all uses of this and let the redesign work fix it in CSS

mankyKitty replied:
Removed `text " "` and added a ticket to fix this with the correct CSS.

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AddAccount.hs:
```diff
@@ -0,0 +1,93 @@
+  => AccountName
+  -> Event t ()
+  -> m (mConf, Event t ())
+uiWalletOnlyAccountCreated newAccount _onClose = do
```
Is `_onClose` placeholder or just making the types agree with `ModalImpl` ?

tomsmalley replied:
type alignment

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AddAccount.hs:
```diff
@@ -0,0 +1,93 @@
+
+  modalFooter $ do
+    onConfirm <- confirmButton def "Done"
+    pure $ (mempty, leftmost [onClose, onConfirm])
```
It seems like this function doesn't actually care at all about `mconf`. Is the `mempty` a placeholder?

mankyKitty replied:
Both this and the `onClose` are part of the `ModalImpl` types.

alexfmpe replied:
Yeah I'm just wondering whether the type-massaging belongs in this function or in the caller

tomsmalley commented on frontend/src/Frontend/ReplGhcjs.hs:
```diff
@@ -93,16 +93,16 @@ app sidebarExtra appCfg = void . mfix $ \ cfg -> do
-    walletCfg' <- mkPage (FrontendRoute_Wallet :/ ()) "wallet" $ do
-      uiWallet (_ide_wallet ideL)
+    (walletCfg', mconf) <- mkPage (FrontendRoute_Wallet :/ ()) "wallet" $ do
+      uiWallet (_ide_wallet ideL) ideL
```
No point in the first argument now

tomsmalley commented on frontend/src/Frontend/UI/Wallet.hs:
```diff
@@ -70,14 +71,25 @@ type HasUiWalletModelCfg mConf key m t =
+uiWallet w m = divClass "keys group" $ do
+  eOpenAddAccount <- confirmButton def "+ Add Account"
+
+  onCreate <- uiCreateKey w
```
Seems like we no longer need this?

tomsmalley commented on frontend/src/Frontend/UI/Wallet.hs:
```diff
@@ -70,14 +71,25 @@ type HasUiWalletModelCfg mConf key m t =
+  pure ( keysCfg & walletCfg_genKey .~ fmap (\a -> (a, "0", "")) onCreate -- TODO let user pick chain/notes
+       , mempty & modalCfg_setModal .~ leftmost
+         [ Just (uiCreateWalletOnlyAccount m) <$ eOpenAddAccount
+         , Just . uiWalletOnlyAccountCreated <$> (w ^. wallet_walletOnlyAccountCreated)
```
The presence of two dialogs here seems off, I think we should just have one which contains a workflow.

tomsmalley reviewed (changes requested):
~Does this modal not have to send a TX to create the account?~

Commit:
WIP: Remove the use of two modals and use a workflow.

The add wallet only account dialog now more correctly uses a workflow in a modal to handle
the two step.s

Updated various types to suit, this has broken the account delete function and it won't
fire the events.

removed some margin placeholders `text " "` from a few files, making ticket to handle with
proper margins in redesign.

Commit:
Remove debug button

Change exports for AddAccount dialog to provide the button that provides the modal
configuration.

Fix the issue that was breaking the other dialogs from working.

Clear some warnings

alexfmpe commented on frontend/src/Frontend/UI/Dialogs/AddAccount.hs:
```diff
@@ -7,34 +8,53 @@ module Frontend.UI.Dialogs.AddAccount
-uiCreateWalletOnlyAccount model _onClose = do
-  onClose <- modalHeader $ text "Add Account"
+uiCreateWalletOnlyAccount model onClose = mdo
+  onClose <- modalHeader $ dynText title
```
Shadows argument

mankyKitty replied:
Something very weird happened here... The code you commented on did indeed shadow that variable. However my local version did not as the input to the function was named `_onClose` not `onClose`.

Updating this code to rename the variable to `onCloseExternal` now all of a sudden shows the input to the function as having been  `_onClose` and not `onClose`. :/

Commit:
Don't ignore an external close event trigger

tomsmalley commented on frontend/src/Frontend/UI/Dialogs/AddAccount.hs:
```diff
@@ -0,0 +1,121 @@
+       , HasAddAccountModelCfg model mConf key t
+       )
+  => ModalIde m key t
+  -> m (ModalIdeCfg m key t)
```
This should be fully polymorphic

tomsmalley reviewed (changes requested):
I wired this into the wallet UI for testing, but I get a causality loop exception when finishing the workflow after creating the account.

alexfmpe committed:
Generalize signature

alexfmpe committed:
Fix infinite loop when closing modal

tomsmalley committed:
Merge remote-tracking branch 'origin/redesign' into sc-add-account-dialog
